cql3/statements/describe_statement.cc

```diff
@@ -160,14 +161,72 @@ description type(replica::database& db, const lw_shared_ptr<keyspace_metadata>&
     return description(db, *udt, true);
 }
 
-future<std::vector<description>> types(replica::database& db, const lw_shared_ptr<keyspace_metadata>& ks, bool with_stmt = false) {
-    auto user_types_map = ks->user_types().get_all_types();
-    auto user_types = boost::copy_range<std::vector<shared_ptr<const keyspace_element>>>(user_types_map | boost::adaptors::transformed([] (const auto& e) {
-        // return static_pointer_cast<const keyspace_element>(e.second);
-        return e.second;
-    }));
+struct udt_node {
+    user_type udt;
+    std::vector<sstring> used_by; 
+};
+
+static future<> udts_sorting_aux(const sstring& name,
```

Comment:
I find it impossible to review the correctness of this topological sorting algorithm. You split into two patches (a part in cql.cc, a part here) and half a dozen functions. Since you have more than one visible function that may be used in the future by other developers, you should have C++ unit tests for these functions and their interesting edge cases. If you want to avoid that, you can make all the helper functions static except the one main udt sorting function that is used by DESC SCHEMA, and test the correctness of the latter in a CQL test.